distrobuilder/chroot.go

```diff
@@ -4,80 +4,124 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"regexp"
+	"strconv"
 	"syscall"
 
 	"github.com/lxc/distrobuilder/managers"
 	"github.com/lxc/distrobuilder/shared"
 )
 
-func prepareChroot(cacheDir string) ([]string, error) {
-	type Mount struct {
-		source string
-		target string
-		fstype string
-		flags  uintptr
-		data   string
+type chrootMount struct {
+	source  string
+	target  string
+	fstype  string
+	flags   uintptr
+	data    string
+	remove  bool
+	mounted bool
+	isDir   bool
+}
+
+func mountFilesystems(rootfs string, mounts []chrootMount) bool {
+	ok := true
+
+	for i, mount := range mounts {
+		if mount.isDir {
+			os.MkdirAll(filepath.Join(rootfs, mount.target), 0755)
+		} else {
+			os.Create(filepath.Join(rootfs, mount.target))
+		}
+		err := syscall.Mount(mount.source, filepath.Join(rootfs, mount.target),
+			mount.fstype, mount.flags, mount.data)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to mount '%s': %s\n", mount.source, err)
+			ok = false
+			break
+		}
+		mounts[i].mounted = true
 	}
 
-	var (
-		err      error
-		unmounts []string
-	)
+	return ok
+}
 
-	mounts := []Mount{
-		{filepath.Join(cacheDir, "rootfs"), "", "tmpfs", syscall.MS_BIND, ""},
-		{"proc", "proc", "proc", 0, ""},
-		{"sys", "sys", "sysfs", 0, ""},
-		{"udev", "dev", "devtmpfs", 0, ""},
-		{"shm", "/dev/shm", "tmpfs", 0, ""},
-		{"/dev/pts", "/dev/pts", "tmpfs", syscall.MS_BIND, ""},
-		{"run", "/run", "tmpfs", 0, ""},
-		{"tmp", "/tmp", "tmpfs", 0, ""},
+func unmountFilesystems(rootfs string, mounts []chrootMount) {
+	// unmount targets in reversed order
+	for i := len(mounts) - 1; i >= 0; i-- {
+		if mounts[i].mounted {
+			err := syscall.Unmount(filepath.Join(rootfs, mounts[i].target), 0)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Failed to unmount '%s': %s\n", mounts[i].target, err)
+				continue
+			}
+			mounts[i].mounted = false
+
+			if mounts[i].remove {
+				os.RemoveAll(filepath.Join(rootfs, mounts[i].target))
+			}
+		}
 	}
+}
 
-	os.MkdirAll(filepath.Join(cacheDir, "rootfs", "dev", "pts"), 0755)
+func killChrootProcesses(rootfs string) error {
```

Comment:
Hm, so can you clarify the intent of this function for me once again, please?

Reply:
So what I thought is that in your chroot you are in a new pid namespace (But I haven't looked that closely at the code.). Which essentially means that you could simply kill `PID 1` in your pid namespace which would automatically cause the kernel to SIGKILL every other process in there. Additionally, the way you're doing it right now is racy in so far that you're succeptible to fork bombs but that might not really be an issue and just be part of my normal paranoia.

Reply:
Another thing that comes to my mind is that if you are worried about lingering process after the parent of the chroot process has died you could think about using the PDEATH signal via prctl. You'd basically set the death signal for the child process if the parent dies. In your case you could simply set it to SIGKILL which would cause the kernel to SIGKILL the child when the parent dies.

Reply:
I tried just killing `PID 1`, but the spawned processes don't die. For that reason, after we've exited the chroot, we go through the processes and kill the ones we need to.

Reply:
Can you paste me that code please, I'll also take a look at your pid namespace usage now.

Reply:
Ah, I see what the problem is:
- your whole app runs in a separate pid namespace, i.e. you don't have a parent process that is running inside of another pid namespace than the child process. This is how I imagined it should be done but since you've talked with @stgraber about this i guess you had a reason for doing it this way.
- you can't sigkill pid1 in a new pid namespace from within that namespace. The kernel will enforce standard init semantics in the pid namespace.

Reply:
The way I thought this might work is to have the main go program call a command via `RunCommand()` and this subcommand has a `constructor` attribute and places the specific distrobuild process into a separate pid and mount namespace which does all of the heavy lifting. If then something goes wrong the main process can SIGKILL the PID that it received thereby tearing down all lingering process by  relying on kernel pid namespace semantics. Additionally the child could set a pdeath signal to SIGKILL such that when the parent dies the child automatically gets killed by the kernel as well. But that is a different design then the one you had. In essence this would be:

```
1. main.go:

    RunCommand("builder")

2. builder.go

   /* C runtime */
    __constructor__ (
            /* set pdeath signal */
            prctl(PR_SET_PDEATHSIG, SIGKILL, 0, 0, 0)
            unshare(CLONE_NEWNS | CLONE_NEWPID)
   )

   /* go runtime */
   func main() {
           // do everything that you did before in the binary you have now
   }
   ```

Then if something goes wrong or you need to cleanup the parent in `main.go` can just kill the child. Also, you could easily run multiple distrobuilds at once. But as I said, different design.